lib/tapioca/sorbet_ext/generic_name_patch.rb

```diff
@@ -60,60 +85,62 @@ def name
         end
       end
 
-      prepend GenericNamePatch
+      prepend GenericPatch
     end
   end
 end
 
 module Tapioca
-  class TypeMember < T::Types::TypeMember
+  # This is subclassing from `Module` so that instances of this type will be modules.
+  # The reason why we want that is because that means those instances will automatically
+  # get bound to the constant names they are assigned to by Ruby. As a result, we don't
+  # need to do any matching of constants to type variables to bind their names, Ruby will
+  # do that automatically for us and we get the `name` method for free from `Module`.
+  class TypeVariableModule < Module
     extend T::Sig
 
-    sig { returns(T.nilable(String)) }
-    attr_accessor :name
-
-    sig { returns(T.untyped) }
-    attr_reader :fixed, :lower, :upper
+    class Type < T::Enum
+      enums do
+        Member = new("type_member")
+        Template = new("type_template")
+      end
+    end
 
-    sig { params(variance: Symbol, fixed: T.untyped, lower: T.untyped, upper: T.untyped).void }
-    def initialize(variance, fixed, lower, upper)
-      super(variance)
+    sig do
+      params(context: Module, type: Type, variance: Symbol, fixed: T.untyped, lower: T.untyped, upper: T.untyped).void
+    end
+    def initialize(context, type, variance, fixed, lower, upper) # rubocop:disable Metrics/ParameterLists
+      @context = context
+      @type = type
+      @variance = variance
       @fixed = fixed
       @lower = lower
       @upper = upper
+      super()
     end
 
     sig { returns(String) }
-    def serialize
-      parts = []
-      parts << ":#{@variance}" unless @variance == :invariant
-      parts << "fixed: #{@fixed}" if @fixed
-      parts << "lower: #{@lower}" unless @lower == T.untyped
-      parts << "upper: #{@upper}" unless @upper == BasicObject
-
-      parameters = parts.join(", ")
-
-      serialized = +"type_member"
-      serialized << "(#{parameters})" unless parameters.empty?
-      serialized
-    end
-  end
-
-  class TypeTemplate < T::Types::TypeTemplate
-    extend T::Sig
-
-    sig { returns(T.nilable(String)) }
-    attr_accessor :name
+    def constant_name
+      constant_name = name
+
+      # This is a hack to work around modules under anonymous modules not having
+      # names in 2.6 and 2.7: https://bugs.ruby-lang.org/issues/14895
+      #
+      # This happens when a type variable is declared under `class << self`, for
+      # example.
+      #
+      # The workaround is to give the parent context a name, at which point, our
+      # module gets bound to a name under that name, as well.
+      unless constant_name
+        constant_name = bind_and_get_name_pre_3_0
+      end
 
-    sig { returns(T.untyped) }
-    attr_reader :fixed, :lower, :upper
+      # If we still don't have a name, fallback to `T.untyped`
+      unless constant_name
+        return "T.untyped"
```

Comment:
This feels weird considering the name of the method `constant_name`. Shouldn't we be returning `nil` and let the client code put `T.untyped` in place of the type variable?

Reply:
Good call, since if we ever fall back to `T.untyped`, we would actually generate bad code like:
```ruby
T.untyped = type_member
```

This really should return `nil`, indeed.